File: RPC/2020:06:13 Junio/editorial/computercache/submissions/accepted/ComputerCache_tgr.py
from operator import itemgetter
class BITSum:
   def __init__(self):
      self.arr = []
      self.sz = 0
   def resize(self, sz):
      self.sz = sz
      while len(self.arr) < 4*sz+1:
         self.arr.append(0)
   def dorange(self, a, b, v):
      while a < b:
         lowb = a & - a
         if lowb == 0:
            lowb = b + b
            while lowb & (lowb - 1):
               lowb &= lowb - 1
         while a + lowb > b:
            lowb >>= 1
         self.arr[2*a+lowb] += v
         a += lowb
   def accum(self, a):
      r = 0
      b = 1
      while b < 2*self.sz:
         r += self.arr[2*a+b]
         a &= ~b
         b *= 2
      return r
class BITMax:
   def __init__(self):
      self.arr = []
      self.sz = 0
   def resize(self, sz):
      self.sz = sz
      while len(self.arr) < 4*sz+1:
         self.arr.append(-1)
   def dorange(self, a, b, v):
      while a < b:
         lowb = a & - a
         if lowb == 0:
            lowb = b + b
            while lowb & (lowb - 1):
               lowb &= lowb - 1
         while a + lowb > b:
            lowb >>= 1
         self.arr[2*a+lowb] = v
         a += lowb
   def accum(self, a):
      r = -1 
      b = 1
      while b < 2*self.sz:
         r = max(r, self.arr[2*a+b])
         a &= ~b
         b *= 2
      return r
(n,m,q) = map(int, input().split())
dat = []
bm = BITMax()
bm.resize(n)
bsv = []
for i in range(m):
   t = list(map(int, input().split()))
   t.pop(0)
   dat.append(t)
   bsv.append(BITSum())
   bsv[i].resize(len(t))
cmds = []
setvals = []
def ints1(s):
   return int(s)-1
for i in range(q):
   t = list(map(ints1, input().split()))
   if t[0] == 0:
      bm.dorange(t[2], t[2]+len(dat[t[1]]), i)
   if t[0] == 1:
      ci = bm.accum(t[1])
      if ci < 0:
         t.append(0)
      else:
         t[1] -= cmds[ci][2]
         setvals.append([ci, i])
   cmds.append(t)
setvals = sorted(setvals, key=itemgetter(0))
svat = 0
for i in range(len(cmds)):
   while svat < len(setvals) and setvals[svat][0] == i:
      ci = setvals[svat][1]
      svat += 1
      di = cmds[i][1]
      add = cmds[ci][1]
      cmds[ci].append((bsv[di].accum(add) + dat[di][add]) % 256)
   if cmds[i][0] == 2:
      bsv[cmds[i][1]].dorange(cmds[i][2], cmds[i][3]+1, 1)
for i in range(len(cmds)):
   if cmds[i][0] == 1:
      print(cmds[i][-1]
)

File: RPC/2020:06:13 Junio/editorial/computercache/submissions/accepted/ComputerCache_tgr.cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std ;
struct quad {
   int cmd, i, p, r ;
} ;
template<typename T> struct BIT {
   BIT<T>(int n=0) { resize(n) ; }
   void resize(int n) {
      sz = n ;
      arr.resize(4*n+1) ;
   }
   void dorange(int a, int b, int v) {
      while (a < b) {
         int lowb = a & -a ;
         if (lowb == 0) {
            lowb = b + b ;
            while (lowb & (lowb - 1))
               lowb &= lowb-1 ;
         }
         while (a + lowb > b)
            lowb >>= 1 ;
         arr[2*a+lowb].update(v) ;
         a += lowb ;
      }
   }
   int accum(int a) {
      int r = T::zero ;
      for (int b=1; b<2*sz; b *= 2) {
         r = T::sum(r, arr[2*a+b]) ;
         a &= ~b ;
      }
      return r ;
   }
   int sz ;
   vector<T> arr ;
} ;
struct bitsum {
   bitsum() : d(zero) {}
   void update(int v) { d += v ; }
   static int sum(int a, const bitsum &b) { return a+b.d ; }
   static const int zero = 0 ;
   int d ;
} ;
struct bitmax {
   bitmax() : d(zero) {}
   void update(int v) { d = v ; }
   static int sum(int a, const bitmax &b) { return max(a, b.d) ; }
   static const int zero = -1 ;
   int d ;
} ;
int main() {
   int n, m, q ;
   cin >> n >> m >> q ;
   vector<vector<int> > dat(m) ;
   BIT<bitmax> bm(n) ;
   vector<BIT<bitsum> > bsv(m) ;
   for (int i=0; i<m; i++) {
      int ts ;
      cin >> ts ;
      bsv[i].resize(ts) ;
      auto &v = dat[i] ;
      v.resize(ts) ;
      for (auto &u : v)
         cin >> u ;
   }
   vector<quad> cmds(q) ;
   vector<pair<int,int> > setvals ;
   for (int i=0; i<(int)cmds.size(); i++) {
      auto &v = cmds[i] ;
      cin >> v.cmd ;
      switch (v.cmd) {
case 1: cin >> v.i >> v.p ; v.i-- ; v.p-- ;
        bm.dorange(v.p, v.p+dat[v.i].size(), i) ;
        break ;
case 2: cin >> v.p ; v.p-- ;
        {
           int ci = bm.accum(v.p) ;
           if (ci < 0) {
              v.r = 0 ;
           } else {
              v.i = cmds[ci].i ;
              v.p -= cmds[ci].p ;
              setvals.push_back({ci, i}) ;
           }
        }
        break ;
case 3: cin >> v.i >> v.p >> v.r ; v.i-- ; v.p-- ; v.r-- ;
        break ;
      }
   }
   sort(setvals.begin(), setvals.end()) ;
   int svat = 0 ;
   for (int i=0; i<(int)cmds.size(); i++) {
      while (svat < (int)setvals.size() && setvals[svat].first == i) {
         int ci = setvals[svat++].second ;
         int di = cmds[i].i ;
         int add = cmds[ci].p ;
         cmds[ci].r = (bsv[di].accum(add) + dat[di][add]) % 256 ;
      }
      auto &v = cmds[i] ;
      if (v.cmd == 3)
         bsv[v.i].dorange(v.p, v.r+1, 1) ;
   }
   for (int i=0; i<(int)cmds.size(); i++)
      if (cmds[i].cmd == 2)
         cout << cmds[i].r << endl ;
}
